### DISCUSSION TITLE
REST API from Prisma Schema

### DISCUSSION THREAD

**Discussion by VinayaSathyanarayana** [prisma/prisma#5758] [Q&A] [answered]
Is there a way to automatically create a REST API end point from Prisma Schema similar to what loopback.io does?
Is is possible to convert prisma schema to loopback schema - so that loopback can be used?

**Comment by ryands17** [comment]
@VinayaSathyanarayana 👋 

I don't think there's any generator that creates a REST API automatically from Prisma schema. Currently there's [TypeGraphQL](https://github.com/MichalLytek/typegraphql-prisma), that creates a GraphQL API so that might be worth looking into as to how it could be modified to create a REST API.

**Comment by sultan1234** [comment]
try amplication.com
this site for generating REST API from prisma

**Comment by yjavaherian** [comment]
you can use trpc-generator and trpc-openapi toghether

**Comment by ymc9** [comment, marked as answer]
In case people are still looking for this, we created a toolkit called [ZenStack](https://zenstack.dev/?utm_campaign=github&utm_medium=organic&utm_content=prisma-restful) which extends Prisma to include access control support and many other things. The modeling of access control policies looks like the following:

```prisma
model Post {
    id        String @id @default(cuid())
    title     String
    published Boolean @default(false)
    author    User @relation(fields: [authorId], references: [id])
    authorId  String

    // author has full access
    @@allow('all', auth() == author)
    // logged-in users can view published posts
    @@allow('read', auth() != null && published)
}
```

The ability to model access control makes it possible to turn the schema into a secure service automatically. The toolkit has a framework-agnostic restful core and framework-specific adapters (currently for Express, Fastify, and Next.js). Installing the services to the framework is mostly a one-liner, e.g., for Express.js:

```ts
import { withPresets } from '@zenstackhq/runtime';
import { ZenStackMiddleware } from '@zenstackhq/server/express';

app.use(
      '/api/openapi',
      ZenStackMiddleware({
          // getSessionUser extracts the current session user from the request, its
          // implementation depends on your auth solution
          getPrisma: (request) => withPresets(prisma, { user: getSessionUser(request) }),
      })
);
```

If you're interested, please check out [this blog post](https://zenstack.dev/blog/openapi?utm_campaign=github&utm_medium=organic&utm_content=prisma-restful).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
